### PR TITLE
Add some test-vectors

### PR DESCRIPTION
**Pull request by pieterlexis**
As discussed on the mailinglist, this is a PR with test vectors. I've tried to explain each set of bytes in the wire format. Let me know if there's anything that needs changing.

### CODE REVIEW

**Review by bemasc — APPROVED** [review]
Looks good to me.  I wonder if it would be clearer to use a little less whitespace in the RFC 3597 examples, to highlight the distinction between uint16, uint8, and bytestring sequences, but that's up to you.

Waiting for another coauthor to confirm before merging.

File: draft-ietf-dnsop-svcb-https.md [hunk 1]
```diff
@@ -1702,6 +1702,151 @@ the authoritative.  Thus, splitting the RR type would double, or in
...
+    \x00\x03                                           # key 3
+    \x00\x02                                           # length: 2 bytes
+    \x00\x35                                           # value
+
+This example has a key that is not registered, its value is unquoted.
+
+    1 foo.example.com key667=hello
+
+    \# 28 (
+    00 10                                              ; priority
```
Comment (bwelling) [inline draft-ietf-dnsop-svcb-https.md:1759]: This should be 00 01, not 00 10.

File: draft-ietf-dnsop-svcb-https.md [hunk 2]
```diff
@@ -1702,6 +1702,151 @@ the authoritative.  Thus, splitting the RR type would double, or in
...
+    \x00\x05                                           # length 5
+    hello                                              # value
+
+This example has a key that is not registered, its value is quoted and contains
+a decimal-escaped character.
+
+    1 foo.example.com key667="hello\210qoo"
+
+    \# 32 (
+    00 10                                              ; priority
```
Comment (bwelling) [inline draft-ietf-dnsop-svcb-https.md:1778]: This should be 00 01, not 00 10.

File: draft-ietf-dnsop-svcb-https.md [hunk 3]
```diff
@@ -1702,6 +1702,151 @@ the authoritative.  Thus, splitting the RR type would double, or in
...
+    \x02\x9b                                           # key 667
+    \x00\x09                                           # length 9
+    hello\xd2qoo                                       # value
+
+Here, two IPv6 hints are quoted in the presentation format.
+
+    1 foo.example.com ipv6hint="2001:db8::1,2001:db8::53:1"
+
+    \# 57 (
+    00 10                                              ; priority
```
Comment (bwelling) [inline draft-ietf-dnsop-svcb-https.md:1796]: This should be 00 01, not 00 10.

File: draft-ietf-dnsop-svcb-https.md [hunk 4]
```diff
@@ -1702,6 +1702,151 @@ the authoritative.  Thus, splitting the RR type would double, or in
...
+         \x00\x00\x00\x00\x00\x00\x00\x01              # first address
+    \x20\x01\x0d\xb8\x00\x00\x00\x00
+         \x00\x00\x00\x00\x00\x53\x00\x01              # second address
+
+In the next vector, the SvcParamValues are not sorted in presentation format,
+but are sorted in the wire-format.
+
+    16 foo.example.org alpn=h2,h3-19 mandatory=alpn ipv4hint=192.0.2.1
+
+    \# 48 (
```
Comment (bwelling) [inline draft-ietf-dnsop-svcb-https.md:1818]: This should be \\# 46, not \\# 48

File: draft-ietf-dnsop-svcb-https.md [hunk 5]
```diff
@@ -1702,6 +1702,151 @@ the authoritative.  Thus, splitting the RR type would double, or in
...
+    \x03foo\x07example\x03com\x00                      # target
+    \x02\x9b                                           # key 667
+    \x00\x09                                           # length 9
+    hello\xd2qoo                                       # value
+
+Here, two IPv6 hints are quoted in the presentation format.
+
+    1 foo.example.com ipv6hint="2001:db8::1,2001:db8::53:1"
+
+    \# 57 (
```
Comment (bwelling) [inline draft-ietf-dnsop-svcb-https.md:1795]: This should be \\# 55, not \\# 57

**Review by bwelling — COMMENTED** [review]
I just tried testing the RFC 3597 for of these test vectors, and found a number of issues.

**Comment by marka63** [comment]
Needed an escaped comma alpn example.

**Comment by pieterlexis** [comment]
> Needed an escaped comma alpn example.

I'll add that as well

@bwelling thanks! I had to hand-craft these values :). I'll fix them

**Comment by marka63** [comment]
add out-of-order mandatory example

**Comment by pieterlexis** [comment]
> add out-of-order mandatory example

I modified an existing vector to showcase this.

**Comment by bemasc** [comment]
@pieterlexis Thanks.  I guess we might as well add an ALPN with an embedded backslash to that list...

**Comment by pieterlexis** [comment]
> @pieterlexis Thanks. I guess we might as well add an ALPN with an embedded backslash to that list...

Done, please check, this was artisinally handcrafted :) (and found an issue in the PowerDNS text writer)

**Comment by bwelling** [comment]
If there are examples of escaping in ALPN parameters, it might also be good to have examples of escaping in the unregistered key (keyNN) format.

**Comment by pieterlexis** [comment]
> If there are examples of escaping in ALPN parameters, it might also be good to have examples of escaping in the unregistered key (keyNN) format.

The section on mandatory says "To enable simpler parsing, this SvcParamValue MUST NOT contain escape sequences."

**Comment by bwelling** [comment]
> > If there are examples of escaping in ALPN parameters, it might also be good to have examples of escaping in the unregistered key (keyNN) format.
> 
> The section on mandatory says "To enable simpler parsing, this SvcParamValue MUST NOT contain escape sequences."

The text for all of the defined formats other than ALPN says that, but the text for the unregistered key format (which is what I was suggesting an example for) does not.  It specifically says that the value is a char-string, whose definition (as far as I can tell) includes escaping.

**Comment by bemasc** [comment]
@bwelling correct.  For example, you can spell `port=123` as `key3=\000\123`.

File: draft-ietf-dnsop-svcb-https.md [hunk 6]
```diff
@@ -1702,6 +1702,178 @@ the authoritative.  Thus, splitting the RR type would double, or in
...
+    )
+
+    \x00\x01                                           # priority
+    \x03foo\x07example\x03com\x00                      # target
+    \x00\x06                                           # key 6
+    \x00\x20                                           # length 32
+    \x20\x01\x0d\xb8\x00\x00\x00\x00
+         \x00\x00\x00\x00\x00\x00\x00\x01              # first address
+    \x20\x01\x0d\xb8\x00\x00\x00\x00
+         \x00\x00\x00\x00\x00\x53\x00\x01              # second address
```
Comment (wtoorop) [inline draft-ietf-dnsop-svcb-https.md:1811]:
```suggestion
    1 foo.example.com ipv6hint="2001:db8::1,2001:db8:ffff:ffff:ffff:ffff:203.0.113.255"

    \# 55 (
    00 01                                              ; priority
    03 66 6f 6f 07 65 78 61 6d 70 6c 65 03 63 6f 6d 00 ; target
    00 06                                              ; key 6
    00 20                                              ; length 32
    20 01 0d b8 00 00 00 00 00 00 00 00 00 00 00 01    ; first address
    20 01 0d b8 ff ff ff ff ff ff ff ff cb 00 71 ff    ; second address
    )

    \x00\x01                                           # priority
    \x03foo\x07example\x03com\x00                      # target
    \x00\x06                                           # key 6
    \x00\x20                                           # length 32
    \x20\x01\x0d\xb8\x00\x00\x00\x00
         \x00\x00\x00\x00\x00\x00\x00\x01              # first address
    \x20\x01\x0d\xb8\xff\xff\xff\xff
         \xff\xff\xff\xff\xcb\x00\x71\xff              # second address
```

A suggestion for IPv4 mapped IPv6 that also crosses the 39 characters boundry of IPv6 addresses represented as string that would be the max string size if IPv4 mapped IPv6 was not anticipated.

Comment (bemasc) [reply draft-ietf-dnsop-svcb-https.md:1811]: Works for me (as long as that line isn't too wide for TXT format).

Comment (wtoorop) [reply draft-ietf-dnsop-svcb-https.md:1811]: Ouch...  that is unfortunately not the case as TXT format has only a width of 69 characters. The example above has 83.

**Comment by wtoorop** [comment]
FYI, we're planning to create a pull request including the commits in this one, with some failure cases added

**Comment by bemasc** [comment]
@pieterlexis Should we close this in favor of #312?

**Pull request closed without merging**